slick/src/main/scala/slick/basic/BasicBackend.scala

```diff
@@ -296,68 +296,85 @@ trait BasicBackend { self =>
+
+                  if(ctx.cancelled) {
+                    if(ctx.deferredError ne null) throw ctx.deferredError
+                    if(state ne null) { // streaming cancelled before finishing
+                      val oldState = state
+                      state = null
+                      a.cancelStream(ctx, oldState)
+                    }
+                  } else if((realDemand > 0 || (state eq null))) {
+                    val oldState = state
+                    state = null
+                    state = a.emitStream(ctx, realDemand, oldState)
+                  }
+
+                  if(state eq null) { // streaming finished and cleaned up
+                    releaseSession(ctx, true)
+                    ctx.streamingResultPromise.trySuccess(null)
+                  }
+
+                } catch { case NonFatal(ex) =>
+                  if(state ne null) try a.cancelStream(ctx, state) catch ignoreFollowOnError
+                  releaseSession(ctx, true)
+                  throw ex
+
+                } finally {
+                  ctx.streamState = state
+                  if (!ctx.isPinned && ctx.priority(continuation) != WithConnection) connectionReleased = true
+                  ctx.sync = 0
                 }
-              } else if((realDemand > 0 || (state eq null))) {
-                val oldState = state
-                state = null
-                state = a.emitStream(ctx, realDemand, oldState)
-              }
-              if(state eq null) { // streaming finished and cleaned up
-                releaseSession(ctx, true)
-                ctx.streamingResultPromise.trySuccess(null)
+
+                if(debug) {
+                  if(state eq null) streamLogger.debug(s"Sent up to ${str(realDemand)} elements - Stream " + (if(ctx.cancelled) "cancelled" else "completely delivered"))
+                  else streamLogger.debug(s"Sent ${str(realDemand)} elements, more available - Performing atomic state transition")
+                }
+
+                demand = ctx.delivered(demand)
+                realDemand = if(demand < 0) demand - Long.MinValue else demand
+
+              } while ((state ne null) && realDemand > 0)
+
+              if(debug) {
+                if(state ne null) streamLogger.debug("Suspending streaming action with continuation (more data available)")
+                else streamLogger.debug("Finished streaming action")
               }
-            } catch { case NonFatal(ex) =>
-              if(state ne null) try a.cancelStream(ctx, state) catch ignoreFollowOnError
-              releaseSession(ctx, true)
-              throw ex
+
+            } catch {
+              case NonFatal(ex) => ctx.streamingResultPromise.tryFailure(ex)
             } finally {
-              ctx.streamState = state
               if (!ctx.isPinned && ctx.priority(continuation) != WithConnection) connectionReleased = true
```

Comment:
Here's the real fix. We must make sure that the runnable is marked with "connectionReleased = true".